_source/_posts/2022-10-11-angular-dependency-injection.md

```diff
@@ -0,0 +1,332 @@
+
+## Angular's Dependency Injection system
+
+When you use the Angular CLI to generate a service, it automatically adds the code to register the service within Angular's DI system for us. Automation sure is sweet! Services contain business logic code that we want to keep separate from view logic.
+
+When Angular CLI generates a service, it adds an `@Injectable()` TypeScript decorator, which is the bit of code that registers a service within Angular DI system:
+
+```ts
+@Injectable({
+  providedIn: 'root'
+})
+export class MyService {
+}
+```
+
+Without doing anything else, we can use our dependency in the application by injecting it into the consuming code as a constructor parameter:
+
+```ts
+@Component({
+  // standard component metadata here
+})
+export class MyComponent {
+  constructor(private myService: MyService) { }
+}
+```
+
+> In Angular v14, you have a new option to use the `inject()` function instead of injecting the service into the consumer as a constructor parameter.
+
+Another way to register dependencies is to provide them manually through the `providers` array. Different Angular building blocks accept `providers` in the metadata. So you can register a provider like this:
+
+```ts
+@NgModule({
+  imports: // stuff here
+  declarations: // stuff here
+  providers: [
+    MyService
+  ]
+})
+export class AppModule {
+}
+```
+
+There's something else to note, though. Angular's DI system allows you to provide a dependency to different places within the application. We saw an example of this in the first code snippet of the `@Injectable()` TypeScript decorator. Angular CLI automatically generates
+
+```ts
+@Injectable({
+  providedIn: 'root'
+})
+export class MyService {
+}
+```
+
+The configuration option `providedIn: 'root'` specifies where within the application to provide the service. In this case, we're saying provide to "root", which means the root of the app so this service is available across the entirety of the application.
+
+ While having this level of configurability sounds unnecessarily complicated, it allows you to fine-tune which dependency to use in your consuming code. You can configure the `providers` array in other modules and Angular building blocks, such as components and directives. You'll get the provider you configured closest to the consuming code when you do so.
+
+Now that we have a quick overview of how to provide dependencies, let's review an integral piece of Angular's DI system, injection tokens.
+
+## Injection tokens in Angular
+
+Injection tokens allow us to have values and objects as dependencies. This means we can depend on a string, such as "Hello world!", and objects, which include configuration objects and global variables such as Web APIs. But injection tokens are even more remarkable because we can also create dependencies to constructs that don't have a runtime definition, such as interfaces! Let's take a look at an example using an injection token.
```

Comment:
```suggestion
Injection tokens allow us to have values and objects as dependencies. This means we can depend on strings, such as "Hello world!", and objects, which include configuration objects and global variables such as Web APIs. But injection tokens are even more remarkable because we can also create dependencies to constructs that don't have a runtime definition, such as interfaces! Let's take a look at an example using an injection token.
```
I think you're saying "we can depend on strings and objects" right? If so, it's clearer to the reader if both are plural... If I've misunderstood, I guess that proves my point ;-)